File: src/main/java/dijkstra/Dijkstra.kt
package dijkstra

fun main() {
    var graph = mutableMapOf<String, MutableMap<String, Int>>()
    graph["start"] = HashMap()
    graph["start"]!!["a"] = 6
    graph["start"]!!["b"] = 2
    graph["a"] = HashMap()
    graph["a"]!!["fin"] = 1
    graph["b"] = HashMap()
    graph["b"]!!["a"] = 3
    graph["b"]!!["fin"] = 5
    graph["fin"] = HashMap()

    val infinity = Int.MAX_VALUE

    var costs = mutableMapOf<String, Int>()
    costs["a"] = 6
    costs["b"] = 2
    costs["fin"] = infinity

    var parents = mutableMapOf<String, String?>()
    parents["a"] = "start"
    parents["b"] = "start"
    parents["fin"] = null

    var processed = mutableListOf<String>()

    dijkstra(graph, costs, parents, processed)

    parents.forEach { o -> println("${o.key} ${o.value}") }
}

fun dijkstra(
    graph: MutableMap<String, MutableMap<String, Int>>,
    costs: MutableMap<String, Int>,
    parents: MutableMap<String, String?>,
    processed: MutableList<String>
) {
    var node = findTheLowestCostNode(costs, processed)
    while (node != null) {
        val cost = costs[node]
        val neighbors = graph[node]
        for (n in neighbors!!.keys) {
            val newCost = cost?.plus(neighbors[n]!!)
            if (costs[n]!! > newCost!!) {
                costs[n] = newCost
                parents[n] = node
            }
        }
        processed.add(node)
        node = findTheLowestCostNode(costs, processed)
    }
}

fun findTheLowestCostNode(
    inputCosts: MutableMap<String, Int>,
    processed: MutableList<String>
): String? {
    var lowestCost = Int.MAX_VALUE
    var lowestCostNode: String? = null
    for (node in inputCosts.keys) {
        if (inputCosts[node]!! < lowestCost && !processed.contains(node)) {
            lowestCost = inputCosts[node]!!
            lowestCostNode = node
        }
    }
    return lowestCostNode
}

File: src/main/java/dijkstra/dijkstra.py
graph = {"start": {}}
graph["start"]["a"] = 6
graph["start"]["b"] = 2
print(graph.keys())
print(graph.values())

graph["a"] = {}
graph["a"]["fin"] = 1
graph["b"] = {}
graph["b"]["a"] = 3
graph["b"]["fin"] = 5
graph["fin"] = {}

infinity = float("inf")
costs = {"a": 6, "b": 2, "fin": infinity}

parents = {"a": "start", "b": "start", "fin": None}

processed = []


def find_the_lowest_cost_node(input_costs):
    lowest_cost = float("inf")
    lowest_cost_node = None
    for nod in input_costs.keys():
        cst = input_costs[nod]
        if cst < lowest_cost and nod not in processed:
            lowest_cost = cst
            lowest_cost_node = nod
    return lowest_cost_node


node = find_the_lowest_cost_node(costs)
while node is not None:
    cost = costs[node]
    neighbors = graph[node]
    for n in neighbors.keys():
        new_cost = cost + neighbors[n]
        if costs[n] > new_cost:
            costs[n] = new_cost
            parents[n] = node
    processed.append(node)
    node = find_the_lowest_cost_node(costs)

print(costs)
print(parents)
